xdem/volume.py

```diff
@@ -0,0 +1,67 @@
+from __future__ import annotations
+
+import numpy as np
+import pandas as pd
+
+
+def hypsometric_binning(ddem: np.ndarray, dem: np.ndarray, bin_size=50,
+                        normalized_bin_size: bool = False) -> pd.DataFrame:
+    """
+    Separate the dDEM in discrete elevation bins.
+
+    :param ddem: The dDEM as a 2D or 1D array.
+    :param dem: The reference DEM as a 2D or 1D array.
+    :param bin_size: The bin interval size in georeferenced units (or percent; 0-100, if normalized_bin_size=True)
+    :param normalized_bin_size: If the given bin size should be parsed as a percentage of the glacier's elevation range.
+
+    :returns: A Pandas DataFrame with elevation bins and dDEM statistics.
+    """
+
+    assert ddem.shape == dem.shape
+    # Remove all nans, and flatten the inputs.
+    nan_mask = np.isnan(ddem) | np.isnan(dem)
+    ddem = ddem[~nan_mask]
+    dem = dem[~nan_mask]
+
+    # Calculate the mean representative elevations between the two DEMs
+    mean_dem = dem - (ddem / 2)
+
+    # If the bin size should be seen as a percentage.
+    if normalized_bin_size:
+        assert bin_size > 0 and bin_size < 100
+
+        # Get the statistical elevation range to normalize the bin size with
+        elevation_range = np.percentile(mean_dem, 99) - np.percentile(mean_dem, 1)
```

Comment:
Should the 1/99 be hard-coded?

Reply:
It's just for calculating the bin sizes if a fraction is given. I think it's okay to use 1-99. Don't you? The bin sizes will only change slightly depending on the numbers.

Reply:
Agreed

Reply:
Ok, yes I didn't realize it had just little influence. So it's fine having it fixed.
But if I'm not mistaken, currently, ther is not the option to force bins of equal sizes. Instead of using a constant bin_size, the elevation steps could be calculated as percentiles e.g. 5, 10, 15, ... 90, 95%.

Reply:
@adehecq it depends on if `normalized_bin_size` is `True` or `False`. If it's false, `bin_size=50` would be parsed as "one bin every 50 metres".

I am open to naming suggestions to make this clearer.

Reply:
Yes, I totally understand the case where normalized_bin_size=False. But if it's True, then the bin_size is still calculated as a fixed elevation size, not a fixed sample size.

Reply:
Well, you can calculate a percentile bin, then use it in your percentile functions:
`pbins = np.arange(min, max, bin_size)`
`bins = np.percentile(mean_dem, pbins)`

Reply:
[I added this as an issue.](https://github.com/GlacioHack/xdem/issues/39)